tket/src/WeightSubgrMono/EndToEndWrappers/MainSolver.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/EndToEndWrappers/MainSolver.hpp"
+
+#include <algorithm>
+#include <array>
+#include <chrono>
+#include <sstream>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+#include "WeightSubgrMono/EndToEndWrappers/CheckedWeightBounds.hpp"
+#include "WeightSubgrMono/Searching/FixedData.hpp"
+#include "WeightSubgrMono/Searching/SearchBranch.hpp"
+#include "WeightSubgrMono/Searching/SharedData.hpp"
+#include "WeightSubgrMono/Searching/ValueOrdering.hpp"
+#include "WeightSubgrMono/Searching/VariableOrdering.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+typedef std::chrono::steady_clock Clock;
+
+const SolutionStatistics& MainSolver::solve(long long timeout_ms) {
+  MainSolverParameters parameters;
+  parameters.timeout_ms = timeout_ms;
+  return solve(parameters);
+}
+
+const SolutionStatistics& MainSolver::solve(
+    const GraphEdgeWeights& pattern_edges, const GraphEdgeWeights& target_edges,
+    long long timeout_ms) {
+  MainSolverParameters parameters;
+  parameters.timeout_ms = timeout_ms;
+  return solve(pattern_edges, target_edges, parameters);
+}
+
+const SolutionStatistics& MainSolver::solve(
+    const GraphEdgeWeights& pattern_edges, const GraphEdgeWeights& target_edges,
+    const MainSolverParameters& parameters) {
+  initialise(pattern_edges, target_edges);
+  return solve(parameters);
+}
+
+const SolutionStatistics& MainSolver::get_solution_statistics() const {
+  return m_data.statistics;
+}
+
+const SolutionStatistics& MainSolver::initialise(
+    const GraphEdgeWeights& pattern_edges,
+    const GraphEdgeWeights& target_edges) {
+  const auto init_start = Clock::now();
+  const auto init_result = m_data.initialise(pattern_edges, target_edges);
+
+  if (init_result == ReductionResult::FAILURE) {
+    // No solution is possible.
+    m_data.shared_data_ptr.reset();
+    m_data.statistics.finished = true;
+  }
+
+  if (init_result == ReductionResult::FINISHED) {
+    // The initial domains and vertex filtering have led to
+    // a unique solution, which is valid; and stored in the shared data.
+    m_data.statistics.finished = true;
+    TKET_ASSERT(m_data.shared_data_ptr);
+
+    // TODO: Check for int overflow and warn...although doesn't actually matter,
```

Comment:
I would say so; if we don't check and an overflow occurs, we'll get very poor returned solutions (or maybe miss solutions because of weight nogood detection), ultimately giving very bad placement solutions.

Edge weights can get quite big if we're not careful, and we are multiplying two weights.

Note that they are quick one-off checks at the beginning of a problem.